WalletConnectSharp.Sign/Controllers/AddressProvider.cs

```diff
@@ -0,0 +1,185 @@
+                return;
+            }
+
+            // DefaultNamespace is null or not found in RequiredNamespaces, update it
+            DefaultNamespace = DefaultSession.RequiredNamespaces.OrderedKeys.FirstOrDefault();
+            if (DefaultNamespace != null)
+            {
+                DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
+            }
+            else
+            {
+                // TODO The Keys property is unordered! Maybe this needs to be updated
+                DefaultNamespace = DefaultSession.RequiredNamespaces.Keys.FirstOrDefault();
+                if (DefaultNamespace != null)
+                {
+                    DefaultChain = DefaultSession.RequiredNamespaces[DefaultNamespace].Chains[0];
+                }
+            }
+        }
+        else
+        {
+            DefaultNamespace = null;
+        }
+    }
+
+    public Caip25Address CurrentAddress(string @namespace = null, SessionStruct session = default)
+    {
+        @namespace ??= DefaultNamespace;
+        if (string.IsNullOrWhiteSpace(session.Topic)) // default
+            session = DefaultSession;
+
+        // double check
+        if (@namespace == null)
+            throw new ArgumentException("CurrentAddress: @namespace is null");
+        if (string.IsNullOrWhiteSpace(session.Topic))
+            throw new ArgumentException("CurrentAddress: Session is undefined");
+        
+        var defaultNamespace = session.Namespaces[@namespace];
+
+        if (defaultNamespace.Accounts.Length == 0)
+            return null; //The namespace {@namespace} has no addresses connected")
+
+        var fullAddress = defaultNamespace.Accounts[0];
+        var addressParts = fullAddress.Split(":");
+
+        var address = addressParts[2];
+        var chainId = string.Join(':', addressParts.Take(2));
+
+        return new Caip25Address()
+        {
+            Address = address,
+            ChainId = chainId,
+        };
+    }
+
+    public Caip25Address[] AllAddresses(string @namespace = null, SessionStruct session = default)
+    {
+        @namespace ??= DefaultNamespace;
+        if (string.IsNullOrWhiteSpace(session.Topic)) // default
+            session = DefaultSession;
+
+        // double check
+        if (@namespace == null)
+            throw new ArgumentException("CurrentAddress: @namespace is null");
+        if (string.IsNullOrWhiteSpace(session.Topic))
+            throw new ArgumentException("CurrentAddress: Session is undefined");
+        
+        var defaultNamespace = session.Namespaces[@namespace];
+
+        if (defaultNamespace.Accounts.Length == 0)
+            return null; //The namespace {@namespace} has no addresses connected")
+
+        return defaultNamespace.Accounts.Select(addr => new Caip25Address()
+        {
+            Address = addr.Split(":")[2], ChainId = string.Join(":", addr.Split(":").Take(2))
+        }).ToArray();
+    }
```

Comment:
Let's convert the logic from these two methods into an extension for `SessionStruct`? This will allow the retrieval of any session's address.